### PR TITLE
Refs #21442 -- Added JSON parsing for request.data property.

### PR DESCRIPTION
**Pull request by udbhavsomani**

### CODE REVIEW

**Review by auvipy — CHANGES_REQUESTED** [review]
i think you should target the main branch for new feature/bug fix.

**Comment by carltongibson** [comment]
This is for this WIP branch. 

Looks good @udbhavsomani. Can you look at the test failures? 
I'll review properly next week. 

Next phase would be the same for multipart requests.

**Comment by udbhavsomani** [comment]
All the postgresql tests have failed. Are you aware of any possible reasons? @carltongibson

**Comment by akshaykedar1996** [comment]
I'm happy to provide more information on the change made in Django's codebase!

In Django, the request object represents an HTTP request that a client has made to a server. The request.data property is used to access the body of the request, which typically contains data sent by the client.

Prior to the change made in ticket #16570, request.data only supported parsing of form data and files. This meant that if a client sent data in JSON format, it was not automatically parsed and could not be accessed directly using request.data.

In ticket #16570, support for parsing JSON data was added to request.data. This change means that if a client sends data in JSON format, it will be automatically parsed and accessible using request.data.

To enable parsing of JSON data, developers can set the content_type of the request to "application/json":

python

import json
from django.http import HttpRequest

data = {'key': 'value'}
json_data = json.dumps(data)

request = HttpRequest()
request.method = 'POST'
request.content_type = 'application/json'
request._body = json_data.encode('utf-8')

In the example above, the request object is created with the content_type set to "application/json". The _body attribute is set to the encoded JSON data, which will be automatically parsed when request.data is accessed.

Overall, the addition of JSON parsing for request.data provides a more flexible and convenient way to access data sent by clients in JSON format, which is a common format for many web APIs.

**Comment by felixxm** [comment]
@akshaykedar1996 Please stop, these kind of comments are not really helpful and generate unnecessary noise for folks who watch this repository.

File: django/http/request.py [hunk 1]
```diff
@@ -390,6 +391,11 @@ def _load_post_and_files(self):
                 QueryDict(self.body, encoding=self._encoding),
                 MultiValueDict(),
             )
+        elif self.content_type == "application/json":
+            self._post, self._files = (
+                QueryDict(urlencode(json.loads(self.body)), encoding=self._encoding),
```
[the remark this answers is not in the document]
Comment (udbhavsomani) [reply django/http/request.py]: I have made the required changes. 
Yes, added the exception handling there as well as a test for it. 
Kindly review! 
Thanks a lot for your inputs!

**Comment by carltongibson** [comment]
Hi @udbhavsomani — I'm moving the development branch to my fork here: 
https://github.com/carltongibson/django/tree/5.0/add-request.data 

I'll close this PR here. Can I ask you to reopen this PR against that branch, and we can work on it there? Thanks!

**Pull request closed without merging**